tiledb/ml/models/_base.py

```diff
@@ -3,36 +3,33 @@
 import platform
 import time
 from abc import ABC, abstractmethod
-from enum import Enum, unique
-from typing import Any, Generic, Mapping, Optional, Tuple, TypeVar
+from typing import Any, Generic, Mapping, Optional, Tuple, TypeVar, Union
 
-import tiledb
+import numpy as np
 
-from ._cloud_utils import get_cloud_uri
+import tiledb
 
-Model = TypeVar("Model")
+from ._cloud_utils import get_cloud_uri, update_file_properties
+from ._specs import (
+    ModelFileProperties,
+    SklearnSpec,
+    TensorboardFileProperties,
+    TensorBoardSpec,
+    TFSpec,
+    TorchSpec,
+)
+
+Artifact = TypeVar("Artifact")
 Meta = Mapping[str, Any]
 Timestamp = Tuple[int, int]
-
-
-@unique
-class ModelFileProperties(Enum):
-    """
-    Enum Class that contains all model array file properties.
-    """
-
-    TILEDB_ML_MODEL_ML_FRAMEWORK = "TILEDB_ML_MODEL_ML_FRAMEWORK"
-    TILEDB_ML_MODEL_ML_FRAMEWORK_VERSION = "TILEDB_ML_MODEL_ML_FRAMEWORK_VERSION"
-    TILEDB_ML_MODEL_STAGE = "TILEDB_ML_MODEL_STAGE"
-    TILEDB_ML_MODEL_PYTHON_VERSION = "TILEDB_ML_MODEL_PYTHON_VERSION"
-    TILEDB_ML_MODEL_PREVIEW = "TILEDB_ML_MODEL_PREVIEW"
+Spec = Union[TFSpec, TorchSpec, SklearnSpec, TensorBoardSpec]
```

Comment:
This union highlights that there's no need for all these *Spec classes, even more so when they don't share a common base class. All that is needed is a single `Spec` (data)class with two fields, `domain_info` and `fields`. Then all the current *Spec classes can be made functions that create a `Spec` and moved to the respective framework module. In fact they don't even have to become standalone functions since they are called only once, just inline them in each `_create_array_internal`.

Reply:
I though of it as well, however I thought that this is a cleaner solution as each `dataclass` has its own 'if' conditions related to each framework in terms of initialisation of these two fields `domain_info` and `fields`, it would put unnecessary load inside the `create_array` functions. Also imo it is easier maintainable this way. What do you think on that?

Reply:
As I wrote I see no point in these *Spec classes:
- They don't make much sense as standalone classes, they are internal helpers tightly coupled to the respective *Artifact classes.
- They don't share a common base / have incompatible constructor.
- The overriden `_create_array` (which is also needlessly split into `_create_array_internal`) is exactly the right place to call the base `_create_array` with the proper `domain_info` and `fields` (even a 2 field `Spec` dataclass is overkill IMO).

Simpler design:
```py
class BaseArtifact:
    def _create_array(self, domain_info, fields):
        ...

class ChildArtifact(BaseArtifact):
    def _create_array(self, **kwargs):
        # determine domain_info, fields from kwargs
        return super()._create_array(domain_info, fields)
```